voluptuous/schema_builder.py

```diff
@@ -334,11 +334,23 @@ def _compile_mapping(self, schema, invalid_msg=None):
 
         def validate_mapping(path, iterable, out):
             required_keys = all_required_keys.copy()
-            # keeps track of all default keys that haven't been filled
-            default_keys = all_default_keys.copy()
+
+            # Build a map of all provided key-value pairs.
+            key_value_map = {}
+            for key, value in iterable:
+                key_value_map[key] = value
+
+            # Insert default values for non-existing keys.
+            for key in all_default_keys:
+                if not isinstance(key.default, Undefined) and \
+                   key.schema not in key_value_map:
+                    # A default value has been specified for this missing
+                    # key, insert it.
+                    key_value_map[key.schema] = key.default()
+
             error = None
             errors = []
-            for key, value in iterable:
+            for key, value in key_value_map.items():
```

Comment:
I'm pretty sure this line is what's causing the test failure. `key_value_map` is not an `OrderedDict` so it fails the ordering guarantee.